feat(curriculum): add availability table lab (#55802)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-availability-table/66b36358ed4f261d64840c24.md

1. You should have a `div` element with the `id` of `legend`. It should contain a `span` element with the text of `Availability` and a `div` element with the `id` of `legend-gradient`.
1. You should give the `#legend-gradient` element a linear gradient that change from `--color0` to `--color5`. Each color value should have two color-stops (expressed in percentage) to make the transition from one color to the following color a hard line.

# --hints--

You should have a `table` element.

```js
assert(!!$('table').length);
```

You should have only one `table` element.

```js
assert($('table').length === 1);
```

Your table should have at least `5` rows.

```js
assert.isAtLeast(document.querySelectorAll('tr')?.length, 5);
```

Your table should have at least `3` columns.

```js
const rows = [...document.querySelectorAll('tr')];
assert(!!rows.length);
rows.forEach((row) => {
    assert.isAtLeast(row.children.length, 3);
})
```

The first row in your table should contain at least three `th` elements.

```js
assert.isAtLeast(document.querySelectorAll('tr:first-child>th').length, 3)
```

> The first row in your table should not contain `td` elements.

Reviewer: This test is always passing.

Author: Now it should fail until a `tr` is added to the page.

const legendGradient = new __helpers.CSSHelp(document).getStyle('#legend-gradient');
assert(!!legendGradient && legendGradient.backgroundImage.includes('linear-gradient('))
```

You should use two color-stops (expressed in percentage) to make the transition from one color to the following color a hard line for your `#legend-gradient`. Remember to use your `--color#` variables.

```js
const legendGradient = new __helpers.CSSHelp(document).getStyle('#legend-gradient');
assert(!!legendGradient && legendGradient.backgroundImage.match(/var\(\s*--color[0-5]\s*\)\s+\d+%\s+\d+%/g).length === 6)
```

# --seed--

## --seed-contents--

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Availability Table</title>
    <link rel="stylesheet" href="styles.css">
</head>

<body>

</body>

</html>
```

```css

```

# --solutions--

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Availability Table</title>
    <link rel="stylesheet" href="styles.css">
</head>

<body>
    <main>

>         </table>

Reviewer: This seems like here by mistake.

---
id: 66b36358ed4f261d64840c24
title: Build an Availability Table
challengeType: 14
dashedName: build-an-availability-table
demoType: onClick
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should have a table with at least three columns and five rows, headers included.

> 1. Cells in the first row should be table headers.

Reviewer 2: ```suggestion
1. Cells in the first row should be table headers containing days of the week.
```

> 1. Cells in the first column should be table headers containing time values and they should have the `class` of `time`.

Reviewer 2: ```suggestion
1. Cells in the first column should be table headers with a `class` of `time` and contain time values as their text.
```

1. All of the data cells should have the `class` of `available-#`, where `#` is a number from `0` to `5` that specifies the number of available people at that time.

> 1. In your `root` selector, you should define CSS variables named `--color#`, where `#` is a number from `0` to `5`, and used them to give a background color to the corresponding `.available-#` elements.

Reviewer 2: ```suggestion
1. In your `root` selector, you should define six CSS variables named `--color#`, where `#` are the numbers `0` to `5`, and give them each a color value. Use them to set the background color of the corresponding `.available-#` elements.
```

1. In your `root` selector, you should define CSS variables named `--solid-border` and `--dashed-border`. You should use them to style the bottom borders of your data cells alternating between solid and dashed borders, depending on the row. Give the rows either the class of `sharp` or `half` to style them.
1. You should have a `div` element with the `id` of `legend`. It should contain a `span` element with the text of `Availability` and a `div` element with the `id` of `legend-gradient`.

> 1. You should give the `#legend-gradient` element a linear gradient that change from `--color0` to `--color5`. Each color value should have two color-stops (expressed in percentage) to make the transition from one color to the following color a hard line.

Reviewer 2: ```suggestion
1. You should give the `#legend-gradient` element a linear gradient that changes between all the colors from `--color0` to `--color5`. Each color value should have two color-stops (expressed in percentage) to make the transition from one color to the following color a hard line.
```

```js
assert.equal(document.querySelector('#legend span')?.innerText, 'Availability');
```

You should have a `div` element with the `id` of `legend-gradient` inside your `#legend`.

```js
assert.equal(document.querySelector('#legend #legend-gradient')?.tagName, 'DIV');
```

You should set the background image of `#legend-gradient` to a linear gradient.

```js
const legendGradient = new __helpers.CSSHelp(document).getStyle('#legend-gradient');
assert(!!legendGradient && legendGradient.backgroundImage.includes('linear-gradient('))
```

You should use two color-stops (expressed in percentage) to make the transition from one color to the following color a hard line for your `#legend-gradient`. Remember to use your `--color#` variables.

```js
const legendGradient = new __helpers.CSSHelp(document).getStyle('#legend-gradient');
assert(!!legendGradient && legendGradient.backgroundImage.match(/var\(\s*--color[0-5]\s*\)\s+\d+%\s+\d+%/g).length === 6)
```

# --seed--

## --seed-contents--

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Availability Table</title>

>     <link rel="stylesheet" href="styles.css">

Reviewer 3: this will need to be removed per our discussion earlier

Reviewer: We should also add a note like this:

```
**Note:** Be sure to link your stylesheet in your HTML and apply your CSS.
```

---
id: 66b36358ed4f261d64840c24
title: Build an Availability Table
challengeType: 14
dashedName: build-an-availability-table
demoType: onClick
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should have a table with at least three columns and five rows, headers included.

> 1. Cells in the first row should be table headers containing days of the week.

Author: I'm not asking to use all the days of the week, so it should be fine as is.

> 1. Cells in the first column should be table headers with a `class` of `time` and contain time values as their text.

Reviewer 4: ```suggestion
1. Cells in the first column should be table headers with a `class` of `time` and should contain time values as their text.
```

1. All of the data cells should have the `class` of `available-#`, where `#` is a number from `0` to `5` that specifies the number of available people at that time.

> 1. In your `root` selector, you should define six CSS variables named `--color#`, where `#` are the numbers `0` to `5`, and give them each a color value. Use them to set the background color of the corresponding `.available-#` elements.

Reviewer 4: ```suggestion
1. In your `:root` selector, you should define six CSS variables named `--color#`, where `#` is a number from `0` to `5`, and assign them each a color value. Use these variables to set the background color of the corresponding `.available-#` elements.
```
